test/linalg.jl

```diff
@@ -62,3 +64,57 @@ end
         end
     end
 end
+
+###### Test jacobian matrix ######
+params(A::SparseMatrixCSC) = params(A.nzval)
+fitparams(A::SparseMatrixCSC, p) = SparseMatrixCSC(A.m, A.n, A.colptr, A.rowval, fitparams(A.nzval, p))
+params(v::DenseArray{<:Real}) = v
+fitparams(v::DenseArray{<:Real}, p) = p
+params(v::DenseArray{<:Complex{T}}) where T = collect(reinterpret(T, v))
+fitparams(v::DenseArray{<:T}, p::DenseArray{T2}) where {T<:Complex,T2} = collect(reinterpret(Complex{T2}, p))
+
+@i function loss(l, j, y::AbstractArray{<:Real}, A, x)
+    imul!(y, A, x, 1.0, 1.0)
+    l += y[j]
+end
+@i function loss(l, j, y::AbstractArray{<:Complex}, A, x)
+    imul!(y, A, x, 1.0, 1.0)
+    if j%2 == 1
+        l += y[div(j-1,2)+1].re
+    else
+        l += y[div(j-1,2)+1].im
+    end
+end
+
+function nilang_mul_jacobian(A, x::AbstractArray)
+    px, pA = params(x), params(A)
+    J = zeros(T, length(px), length(px) + length(pA))
+    for j=1:length(px)
+        _, _, _, _, gA, gx = NiLang.AD.Grad(loss)(Val(1), 0.0, j, zero(x), A, x)
+        J[j,:] = vcat(grad(params(gA)), grad(params(gx)))
+    end
+    return J
+end
+
+function forwarddiff_mul_jacobian(A, x)
+    pA = params(A)
+    px = params(x)
+    pAx = vcat(pA, px)
+    ForwardDiff.jacobian(pAx) do pAx
```

Comment:
I feel confused about the usage of ForwardDiff.jacobian maybe the most closest protype is [here](https://juliadiff.org/ForwardDiff.jl/stable/user/api/#ForwardDiff.jacobian!), but I get Jacobian is stored in pAx. Thanks for your test examples on Jacobian~

Reply:
In the documentation link, you will see
```julia
ForwardDiff.jacobian(f, x::AbstractArray, cfg::JacobianConfig = JacobianConfig(f, x), check=Val{true}())
```

So, the basic usage is `jacobian(f, x)`, where `x` is an array. What you are not familar might be the do-block syntax:
https://docs.julialang.org/en/v1/manual/functions/#Do-Block-Syntax-for-Function-Arguments